### PR TITLE
Use strict types for applicatives

### PR DESCRIPTION
**Pull request by gusty**
This will switch to use the internal technique of ~~tupling~~ wrapping in order to avoid flexible types in overloads for the Applicative dispatcher.

Right now there are 2 issues:

 - Types implementing `IEnumerable<_>` would default to the `seq<_>` instance. This might be arguably desired.

 - Types implementing `IEnumerable<_>` would default to the `seq<_>` instance, even if they implement their own applicative instance. This is not acceptable.

### CODE REVIEW

**Comment by cannorin** [comment]
I tested with https://github.com/fsprojects/FSharpPlus/pull/207/commits/fea4b08e8ddb8ddce36da264a6b53d07b8fe4d8d but it seems to failing to redirect.

```fsharp
Microsoft (R) F# Interactive version 10.2.3 for F# 4.5
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> open FSharpPlus.TypeLits;;
> let v = vector ((fun i -> i + 1), (fun i -> i * 2));;
[<Struct>]
val v : Vector<(int -> int),S<S<Z>>> = [|<fun:v@2>; <fun:v@2-1>|]

> let u = vector (2, 3);;
[<Struct>]
val u : Vector<int,S<S<Z>>> = [|2; 3|]

> v <*> u;;
[<Struct>]
val it : Vector<int,S<S<Z>>> = [|3; 6|]

> open FSharpPlus;;
> v <*> u;;
val it : seq<int> = seq [3; 4; 4; 6]
```

File: src/FSharpPlus/Functor.fs [hunk 1]
```diff
@@ -139,35 +139,32 @@ type Return =
...
+    static member        ``<*>`` (f: Result<_,'E>     , _: ref<Result<_,'E>>         , x: Result<'T,'E>        , [<Optional>]_output: Result<'b,'E>        , [<Optional>]_mthd: Apply) = Result.apply f x : Result<'U,'E>
+    static member        ``<*>`` (f: Choice<_,'E>     , _: ref<Choice<_,'E>>         , x: Choice<'T,'E>        , [<Optional>]_output: Choice<'b,'E>        , [<Optional>]_mthd: Apply) = Choice.apply f x : Choice<'U,'E>
+    static member inline ``<*>`` (KeyValue(a: 'Key, f), _: ref<KeyValuePair<'Key, _>>, KeyValue(b: 'Key, x: 'T), [<Optional>]_output: KeyValuePair<'Key,'U>, [<Optional>]_mthd: Apply) : KeyValuePair<'Key,'U> = KeyValuePair (Plus.Invoke a b, f x)
 
-    static member        ``<*>`` (f: Map<'Key,_>      , x: Map<'Key,'T>         , [<Optional>]_output: Map<'Key,'U>         , [<Optional>]_mthd: Apply) : Map<'Key,'U> = Map (seq {
+    static member        ``<*>`` (f: Map<'Key,_>      , _: ref<Map<'Key,_>>, x: Map<'Key,'T>         , [<Optional>]_output: Map<'Key,'U>         , [<Optional>]_mthd: Apply) : Map<'Key,'U> = Map (seq {
```
Comment (wallymathieu) [inline src/FSharpPlus/Functor.fs]: What does wrapping ref<T> help with?

File: src/FSharpPlus/Functor.fs [hunk 2]
```diff
@@ -176,19 +173,24 @@ type Apply =
...
+            ((^M or ^I1 or ^I2 or ^R) : (static member ``<*>`` : _*_*_*_*_ -> _) input1, ref input1, input2, output, mthd)
         call(Unchecked.defaultof<Apply>, f, x, Unchecked.defaultof<'``Applicative<'U>``>)
 
     static member inline InvokeOnInstance (f: '``Applicative<'T->'U>``) (x: '``Applicative<'T>``) : '``Applicative<'U>`` =
         ((^``Applicative<'T->'U>`` or ^``Applicative<'T>`` or ^``Applicative<'U>``) : (static member (<*>) : _*_ -> _) (f, x))
 
+type Apply with
+    static member inline ``<*>`` (f: '``Monad<'T->'U>``      , _: ref<'``Monad<'T->'U>``>      , x: '``Monad<'T>``      , _output: '``Monad<'U>``      , _mthd:Default2) : '``Monad<'U>`` = Bind.InvokeOnInstance f (fun (x1: 'T->'U) -> Bind.InvokeOnInstance x (fun x2 -> Return.InvokeOnInstance (x1 x2)))
+    static member inline ``<*>`` (f: '``Applicative<'T->'U>``, _: ref<'``Applicative<'T->'U>``>, x: '``Applicative<'T>``, _output: '``Applicative<'U>``, _mthd:Default1) : '``Applicative<'U>`` = Apply.InvokeOnInstance f x
+    static member inline ``<*>`` (_: ^t when ^t : null and ^t: struct, _, _: ^u when ^u : null and ^u: struct, _output: ^r when ^r : null and ^r: struct, _mthd: Default1) = id
```
Comment (wallymathieu) [inline src/FSharpPlus/Functor.fs]: Or is this the main point of the PR?

File: src/FSharpPlus/Functor.fs (hunk 1 above)
Comment (cannorin) [reply src/FSharpPlus/Functor.fs]: yes that's the point of this PR, F# should not have covariance so the custom type with `IE<_>` implemented should not trigger the overload expecting `ref<_ seq>`

Comment (wallymathieu) [reply src/FSharpPlus/Functor.fs]: Aa, thanks @cannorin !

Comment (gusty) [reply src/FSharpPlus/Functor.fs]: Just a quick update. At the moment it succeed with `f <*> x` but not with `f <*> x <*> y`, in the latter case still tried to infer `seq`, but at least now it fails.

**Comment by gusty** [comment]
Another update.
The other techinique that can be used in place or combined with wrapping is to add a higher DefaultX marker to the interface overload, the problem here if I do that is that type inference gets worst when resolving to those overloads and needs type annotation. This is currently a noticeable problem with the monad overload.
I'm coming to the conclusion that if there's no way to solve that type inference problem, a good compromise would be to let interfaces type inference degrade, that problem is preferred to not having external instances being resolved, but it would clearly be a breaking change.

**Comment by gusty** [comment]
This is passing now, but there are some breaking specially in the idiom brackets section.
Also the Apply instance for `KeyValuePair` is removed.

Now, if we put in the balance this regression against not having externally defined applicatives implementing interfaces working decently (namely something along a plain `f <*> x <*> y`) it should be clear that this is priority.

How we rollout this change? Possibly we would have to skip a 1.1 version and move directly to a 2.0

The open question is whether this could be improved in order minimize the regression. I spent a crazy amount of time looking at this, if I add the time accumulated in fighting applicatives it's probably the time required to fix the F# compiler to make them work properly for once.

If I have more time, I will keep trying, but otherwise this would have to be merged.

Thoughts?

**Comment by cannorin** [comment]
It's working now!

```fsharp
Microsoft (R) F# Interactive version 10.2.3 for F# 4.5
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> open FSharpPlus.TypeLits;;
> let v = vector ((fun i -> i + 1), (fun i -> i * 2));;
[<Struct>]
val v : Vector<(int -> int),S<S<Z>>> = [|<fun:v@2>; <fun:v@2-1>|]

> let u = vector (2, 3);;
[<Struct>]
val u : Vector<int,S<S<Z>>> = [|2; 3|]

> v <*> u;;
[<Struct>]
val it : Vector<int,S<S<Z>>> = [|3; 6|]

>  open FSharpPlus;;
> v <*> u;;
[<Struct>]
val it : Vector<int,S<S<Z>>> = [|3; 6|]
```

I suppose very few people seriously uses idiom brackets, so these breaking changes are acceptable.
Since we are introducing big changes such as #185 and #192 and other breaking change #205 I think it is better to move to 2.0 skipping 1.1.

**Comment by gusty** [comment]
Thanks for your feedback. #185 and #192 are not minor things, but at the same time they don't break anything.

Regarding #205 it's not breaking 1.0 since it didn't exists at that time.

So far this is the first serious reason to break 1.0, however I can think of other upcoming breaking changes:

- A new `ListT` monad transformer I have on drafts, that won't be compatible with the existing one.

- A redesign of the computation expression without the "autosense" functionality which users of this library have demonstrated it works only on certain scenarios and create more confusion than other thing.

**Comment by gusty** [comment]
Continued in #569

**Pull request closed without merging**